test/clj/rems/test/migrations/convert_to_dynamic_applications.clj

```diff
@@ -0,0 +1,499 @@
+                                                    :event/actor (:userid event)
+                                                    :application/id (:id application)
+                                                    :application/comment (:comment event)})
+        "autoapprove" (assert false "not implemented") ; TODO: migrate "autoapprove"
+        "return" (applications/add-dynamic-event! {:event/type :application.event/returned
+                                                   :event/time (:time event)
+                                                   :event/actor (:userid event)
+                                                   :application/id (:id application)
+                                                   :application/comment (:comment event)})
+        "review" (applications/add-dynamic-event! {:event/type :application.event/commented
+                                                   :event/time (:time event)
+                                                   :event/actor (:userid event)
+                                                   :application/id (:id application)
+                                                   ;; TODO: request-id doesn't make much sense for these old applications
+                                                   :application/request-id (UUID. 0 0)
+                                                   :application/comment (:comment event)})
+        "third-party-review" (assert false "not implemented") ; TODO: migrate "third-party-review"
+        "review-request" (assert false "not implemented") ; TODO: migrate "review-request"
+        "withdraw" (assert false "not implemented") ; TODO: migrate "withdraw"
+        "close" (assert false "not implemented"))))) ; TODO: migrate "close"
+
+(deftest test-migration
+  (let [applications (applications/get-applications-impl-batch "whatever" {})
+        application-id 10
+        application (->> applications
+                         (filter #(= application-id (:id %)))
+                         first)
+        dynamic-workflows (->> (workflow/get-workflows {})
+                               (filter #(= "workflow/dynamic" (get-in % [:workflow :type]))))
+        new-workflow (first dynamic-workflows)]
+    (assert application)
+    (assert (= 1 (count dynamic-workflows)))
+
+    (println "--- before ---")
+    (pprint application)
+    (migrate-catalogue-items! (:id new-workflow))
+    (migrate-application! 1 (:id new-workflow))
+    (migrate-application! 2 (:id new-workflow))
+    (migrate-application! 3 (:id new-workflow))
+    (migrate-application! 4 (:id new-workflow))
+    (migrate-application! 5 (:id new-workflow))
+    (migrate-application! 8 (:id new-workflow))
+    (migrate-application! 9 (:id new-workflow))
+    (migrate-application! (:id application) (:id new-workflow))
+    (println "--- after ---")
+    (pprint (applications/get-application-state (:id application))))
+
```

Comment:
you could run `rems.validate/validate` here

you could perhaps also add some checks to the validation code if needed

Reply:
`rems.validate/validate` only checks whether `applications/get-application-state` throws an exception. The actual validation happens every time an event is written or read from the database.